File: Backtracking/Permutation2/perm.py
class Solution:
    def doBackTrack(self, ans, idx, nums):
        n = len(nums)

        if idx == n:
            ans.append(nums.copy())
            return

        st = set()

        for i in range(idx, n):
            if nums[i] in st:
                continue

            st.add(nums[i])

            nums[i], nums[idx] = nums[idx], nums[i]

            self.doBackTrack(ans, idx + 1, nums)

            nums[i], nums[idx] = nums[idx], nums[i]

    def permuteUnique(self, nums):
        ans = []
        self.doBackTrack(ans, 0, nums)
        return ans

File: Backtracking/Permutation2/perm.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void doBackTrack(vector<vector<int>>& ans, int idx,vector<int>& nums){
        int n = nums.size();
        if(idx==n){
            ans.push_back(nums);
            return;
        }
        unordered_set<int>st;
        for(int i=idx;i<n;i++){
            if(st.count(nums[i]))continue;
            st.insert(nums[i]);
            swap(nums[i],nums[idx]);
            doBackTrack(ans,idx+1,nums);
            swap(nums[i],nums[idx]);
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        doBackTrack(ans,0,nums);
        return ans;
    }
};